File: 3953-maximum-score-with-co-prime-element/3953-maximum-score-with-co-prime-element.cpp
class Solution {
public:
    int maxScore(vector<int>& nums, int maxVal) {
        int n = nums.size();
        int e = maxVal;
        int i = 0;
        while (i < n)
        {
            if (nums[i] > e)
            {
                e = nums[i];
            }
            i++;
        }
        vector<int> f(e + 1, 0);
        i = 0;
        while (i < n)
        {
            f[nums[i]]++;
            i++;
        }
        vector<int> c(e + 1, 0);
        i = 1;
        while (i <= e)
        {
            int j = i;
            while (j <= e)
            {
                c[i] += f[j];
                j += i;
            }
            i++;
        }
        vector<int> h(e + 1, 0);
        vector<int> g;
        vector<int> p(e + 1, 0);
        h[1] = 1;
        i = 2;
        while (i <= e)
        {
            if (p[i] == 0)
            {
                g.push_back(i);
                h[i] = -1;
            }
            int j = 0;
            int z = g.size();
            while (j < z && i * g[j] <= e)
            {
                p[i * g[j]] = 1;
                if (i % g[j] == 0)
                {
                    h[i * g[j]] = 0;
                    break;
                }
                else
                {
                    h[i * g[j]] = -h[i];
                }
                j++;
            }
            i++;
        }
        vector<int> w(e + 1, 0);
        int d = 1;
        while (d <= e)
        {
            if (h[d] != 0)
            {
                int v = d;
                while (v <= e)
                {
                    w[v] += h[d] * c[d];
                    v += d;
                }
            }
            d++;
        }
        int s = -100005;
        int v = 1;
        while (v <= e)
        {
            int o = n - w[v];
            if (f[v] > 0)
            {
                int r = 0;
                if (v > 1)
                {
                    r = v - o + 1;
                }
                else
                {
                    r = 1;
                }
                if (r > s)
                {
                    s = r;
                }
            }
            if (v <= maxVal)
            {
                int r = 0;
                if (o > 0)
                {
                    r = v - o;
                }
                else
                {
                    r = v - 1;
                }
                if (r > s)
                {
                    s = r;
                }
            }
            v++;
        }
        return s;
    }
};

File: 3953-maximum-score-with-co-prime-element/3953-maximum-score-with-co-prime-element.java
class Solution {
    public int maxScore(int[] nums, int maxVal) {
        int n = nums.length;
        int mx = maxVal;

        for (int x : nums) mx = Math.max(mx, x);

        int[] freq = new int[mx + 1];
        for (int x : nums) freq[x]++;

        int[] cnt = new int[mx + 1];
        for (int d = 1; d <= mx; d++) {
            for (int m = d; m <= mx; m += d) {
                cnt[d] += freq[m];
            }
        }

        int[] mu = new int[mx + 1];
        int[] primes = new int[mx + 1];
        boolean[] comp = new boolean[mx + 1];
        int pc = 0;
        mu[1] = 1;

        for (int i = 2; i <= mx; i++) {
            if (!comp[i]) {
                primes[pc++] = i;
                mu[i] = -1;
            }
            for (int j = 0; j < pc && i * primes[j] <= mx; j++) {
                int p = primes[j];
                comp[i * p] = true;
                if (i % p == 0) {
                    mu[i * p] = 0;
                    break;
                }
                mu[i * p] = -mu[i];
            }
        }

        long ans = Long.MIN_VALUE;

        for (int v = 1; v <= mx; v++) {
            int cop = 0;

            for (int d = 1; d * d <= v; d++) {
                if (v % d != 0) continue;

                cop += mu[d] * cnt[d];

                int e = v / d;
                if (e != d) cop += mu[e] * cnt[e];
            }

            long base = (long) v - n + 1 + cop;

            if (freq[v] > 0) {
                long pen = (v == 1) ? 1 : 0;
                ans = Math.max(ans, base - pen);
            }

            if (v <= maxVal && freq[v] == 0) {
                long pen = (cop == n) ? 2 : 1;
                ans = Math.max(ans, base - pen);
            }
        }

        return (int) ans;
    }
}